File: src/client/java/com/somefrills/features/mining/GemstoneDesyncFix.java
package com.somefrills.features.mining;


import com.google.common.collect.Sets;
import com.somefrills.config.FrillsConfig;
import com.somefrills.events.BlockUpdateEvent;
import com.somefrills.features.core.AreaFeature;
import com.somefrills.misc.Area;
import com.somefrills.misc.Utils;
import meteordevelopment.orbit.EventHandler;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;

import java.util.HashSet;

import static com.somefrills.Main.mc;
import static net.minecraft.block.HorizontalConnectingBlock.*;

public class GemstoneDesyncFix extends AreaFeature {

    private static final HashSet<Area> islands = Sets.newHashSet(
            Area.DWARVEN_MINES,
            Area.CRYSTAL_HOLLOWS,
            Area.MINESHAFT,
            Area.CRIMSON_ISLE,
            Area.THE_RIFT
    );

    public GemstoneDesyncFix() {
        super(FrillsConfig.instance.mining.gemstoneDesyncFixEnabled);
    }

    public static boolean isDefaultPane(BlockState state) {
        return Utils.isStainedGlass(state) && !isConnectedPane(state);
    }

    public static boolean isConnectedPane(BlockState state) {
        return state.get(NORTH) || state.get(EAST) || state.get(SOUTH) || state.get(WEST);
    }

    public static BlockState asFullPane(BlockState state) {
        return state.with(NORTH, true).with(EAST, true).with(SOUTH, true).with(WEST, true);
    }

    @EventHandler
    private void onBlock(BlockUpdateEvent event) {
        if (event.newState.isAir() && Utils.isStainedGlass(event.oldState)) {
            event.newState.updateNeighbors(mc.world, event.pos, Block.NOTIFY_ALL);
        }
    }

    @Override
    protected boolean checkArea(Area area) {
        return islands.contains(area);
    }
}

File: src/main/kotlin/com/somefrills/features/mining/GemstoneDesyncFix.kt
package com.somefrills.features.mining

import com.somefrills.Main.mc
import com.somefrills.config.FrillsMod

import com.somefrills.events.BlockUpdateEvent
import com.somefrills.features.core.AreaFeature
import com.somefrills.modules.FrillsFeature
import com.somefrills.misc.Area
import com.somefrills.utils.isStainedGlass
import com.somefrills.events.core.EventHandle
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.CrossCollisionBlock.*

@FrillsFeature
object GemstoneDesyncFix : AreaFeature(FrillsMod.config.mining.gemstoneDesyncFixEnabled) {
    @EventHandle
    private fun onBlock(event: BlockUpdateEvent) {
        val level = mc.level ?: return
        if (event.newState.isAir && event.oldState.isStainedGlass()) {
            event.newState.updateNeighbourShapes(level, event.pos, Block.UPDATE_ALL);
        }
    }

    @JvmStatic
    fun onGetUpdateState(original: BlockState): BlockState {
        if (isEnabled() && isDefaultPane(original)) {
            return asFullPane(original)
        }
        return original
    }

    private fun isDefaultPane(state: BlockState): Boolean {
        return state.isStainedGlass() && !isConnectedPane(state)
    }

    private fun isConnectedPane(state: BlockState): Boolean {
        return state.getValue(NORTH) || state.getValue(EAST) || state.getValue(SOUTH) || state.getValue(WEST)
    }

    private fun asFullPane(state: BlockState): BlockState {
        return state.setValue(NORTH, true).setValue(EAST, true).setValue(SOUTH, true).setValue(WEST, true)
    }

    override fun checkArea(area: Area): Boolean {
        return area in setOf(
            Area.DWARVEN_MINES,
            Area.CRYSTAL_HOLLOWS,
            Area.MINESHAFT,
            Area.CRIMSON_ISLE,
            Area.THE_RIFT
        )
    }

}